### PR TITLE
Creation of SYCL tutorial on fp reassociate and fp contract pragmas

### PR DESCRIPTION
**Pull request by cassiekexin**
Signed-off-by: cassie.li <kexinli@scc823104.sc.intel.com>

# Description

* This changelist adds a tutorial design including README.md, etc. along  with a regtest testing the tutorial
* CL to enable the two floating point pragmas: https://swarm61.devtools.intel.com:6161/changes/6550362

Fixes # (issue) 

## Type of change

Please delete options that are not relevant. Add a 'X' to the one that is applicable. 

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Sample Migration (Moving sample from old repository after completing checklist established)

# How Has This Been Tested?

I wrote a regtest under regtest/hld/sycl/tutorials/fp_pragmas/...
https://spetc.intel.com/testsummary?testRunIds=5281666

- [X] Command Line
- [ ] oneapi-cli
- [ ] Visual Studio
- [ ] Eclipse IDE
- [ ] VSCode

# Checklist for Moving samples:
Links and Details can be found in the samples WG Teams Files. 

- [ ] Review sample design with domain reviewers https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts 
- [X] Implement coding guidelines and ensure code quality.
- [X] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [X] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com
- [X] Adhere to readme template 
- [X] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [X] Enforce format via clang-format config file
- [X] Compile code using compiler flags and fix anything detected "enable /Wall -Wformat-security -Werror=format-security"
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth or Joe Oster. (GitHub User: JoeOster)
- [ ] Tested using Dev Cloud when applicable
- [ ] Implement fixes for ONSAM Jiras
- [ ] If you have new dependencies/binaries, inform Samples Project Manager Swapna R Dontharaju (@srdontha) or @JoeOster

### CODE REVIEW

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
+#Contraction and Reassociation on Floating Points
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md]: Change to: ...on Floating Point Numbers

I think you need a space after the #

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+The [oneAPI DPC++ FPGA Optimization Guide](https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide) is the reference manual for targeting FPGAs through DPC++. <br>
+The [oneAPI Programming Guide](https://software.intel.com/en-us/oneapi-programming-guide) is a general resource for target-independent DPC++ programming.
+
+| Optimized for                     | Description
+---                                 |---
+| OS                                | Linux* Ubuntu* 18.04; Windows* 10
+| Hardware                          | Intel® Programmable Acceleration Card (PAC) with Intel Arria® 10 GX FPGA; <br> Intel® FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix® 10 SX)
+| Software                          | Intel® oneAPI DPC++ Compiler <br> Intel® FPGA Add-On for oneAPI Base Toolkit
+| What you will learn               | The basic usage of the `fp contract(fast|off)` and `fp reassociate(on|off)` pragmas <br> How the `fp contract(fast|off)` and `fp reassociate(on|off)` pragmas affect resource use <br> How to apply the `fp contract(fast|off)` and `fp reassociate(on|off)` pragmas in your program
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md]: Something is wrong with the markdown here, when viewed in 'rich diff' format (so through a markdown interpreter) this line gets cut off after fp contract(fast|off) - please fix.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+This FPGA tutorial explains how to use the fp reassociate and fp contract pragmas for floating points.
+
+***Documentation***:  The [DPC++ FPGA Code Samples Guide](https://software.intel.com/content/www/us/en/develop/articles/explore-dpcpp-through-intel-fpga-code-samples.html) helps you to navigate the samples and build your knowledge of DPC++ for FPGA. <br>
+The [oneAPI DPC++ FPGA Optimization Guide](https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide) is the reference manual for targeting FPGAs through DPC++. <br>
+The [oneAPI Programming Guide](https://software.intel.com/en-us/oneapi-programming-guide) is a general resource for target-independent DPC++ programming.
+
+| Optimized for                     | Description
+---                                 |---
+| OS                                | Linux* Ubuntu* 18.04; Windows* 10
+| Hardware                          | Intel® Programmable Acceleration Card (PAC) with Intel Arria® 10 GX FPGA; <br> Intel® FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix® 10 SX)
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md]: Please remove the smicolon at the end of the first line

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+| What you will learn               | The basic usage of the `fp contract(fast|off)` and `fp reassociate(on|off)` pragmas <br> How the `fp contract(fast|off)` and `fp reassociate(on|off)` pragmas affect resource use <br> How to apply the `fp contract(fast|off)` and `fp reassociate(on|off)` pragmas in your program
+| Time to complete                  | 20 minutes
+
+
+
+## Purpose
+This tutorial demonstrates a simple example of using the `fp contract(fast|off)` and `fp reassociate(on|off)` pragmas.
+
+### Description of the pragmas
+The `fp contract(fast|off)` pragma controls whether the compiler can skip intermediate rounding and conversions mainly between double precision arithmetic operations. If multiple occurrences of this pragma affect the same scope of your code, the pragma with the narrowest scope takes precedence.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md]: What does 'mainly' mean here?  Does this pragma not apply to single precision arithmetic?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+The `fp reassociate(on|off)` pragma controls the relaxing of the order of floating point arithmetic operations within the code block that this pragma is applied to. If multiple occurrences of this pragma affect the same scope of your code, the pragma with the narrowest scope takes precedence.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md]: I think both of these descriptions should be re-written, I don't find them very clear.  Work with your manager or a senior member on your team if you need some help.

I assume the default for both of these is off?  Which means the purpose of the off pragma is to allow you to define a block of code where the pragma is enabled between a (on / fast) and (off) pragma?  Please make this more clear.

Please provide a bit more detail on what each of the pragmas mean, possibly an example of each?

The part about multiple occurrences is probably not necessary in the tutorial, as long as it's properly documented.  Also I don't understand what this sentence even means.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+#### Example
+This tutorial design applies `fp contract(off)` and `fp reassociate(off)` at a global scope.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md]: Why is this necessary?  Is the default for these to be on?  How do these pragmas interact with command line arguments related to floating point operations?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+```
+
+A kernel in this tutorial applies `fp reassociate(on)` to a sequence of addition. The following is an example:
+
+```
+#pragma clang fp reassociate(on)
+  accessorRes[0] += 
+    accessorA[0] + accessorB[0] + accessorC[0] + accessorD[0];
+```
+
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md]: For both of the examples above, please explain what effect the pragma has on the code.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/sample.json [hunk 8]
```diff
@@ -0,0 +1,55 @@
+{
+  "guid": "7909FAE1-D3D4-4E97-A963-14A884F33495",
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/sample.json]: Please confirm you followed the procedure to generate a unique guid for this tutorial, and did not simply copy and paste this from a different tutorial.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/sample.json [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  "name": "FP Pragmas",
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/sample.json]: Please change to Floating Point Pragmas.  

In fact, I don't really like the directory name 'fp_pragmas', what do you think of changing it to 'floating_point_pragmas'?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md [hunk 10]
```diff
@@ -0,0 +1,178 @@
+#Contraction and Reassociation on Floating Points
+This FPGA tutorial explains how to use the fp reassociate and fp contract pragmas for floating points.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md]: floating point numbers, not floating points.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/src/fp_pragmas.cpp [hunk 11]
```diff
@@ -0,0 +1,259 @@
+//==============================================================
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/src/fp_pragmas.cpp]: Please confirm you ran clang-format on this file?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/src/fp_pragmas.cpp [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+#include <vector>
+
+// dpc_common.hpp can be found in the dev-utilities include folder.
+// e.g., $ONEAPI_ROOT/dev-utilities//include/dpc_common.hpp
+#include "dpc_common.hpp"
+
+using namespace sycl;
+using namespace std;
+
+#pragma clang fp contract(off)
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/src/fp_pragmas.cpp]: Comment here to indicate why you are applying these 'off' pragmas here and what it means for the tutorial.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/src/fp_pragmas.cpp [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+      accessor accessorB(bufferB, h, read_only);
+      accessor accessorC(bufferC, h, read_only);
+      accessor accessorD(bufferD, h, read_only);
+      accessor accessorRes(bufferE, h, write_only, noinit);
+
+      // FPGA-optimized kernel
+      // Using kernel_args_restrict tells the compiler that the input
+      // and output buffers won't alias.
+      h.single_task<ContractFastKernel>([=]() [[intel::kernel_args_restrict]] {
+#pragma clang fp contract(fast)
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/src/fp_pragmas.cpp]: Please add some comments here to explain what the fp contract pragma allows the compiler to do here to save area.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/src/fp_pragmas.cpp [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+      accessor accessorB(bufferB, h, read_only);
+      accessor accessorC(bufferC, h, read_only);
+      accessor accessorD(bufferD, h, read_only);
+      accessor accessorRes(bufferE, h, write_only, noinit);
+
+      // FPGA-optimized kernel
+      // Using kernel_args_restrict tells the compiler that the input
+      // and output buffers won't alias.
+      h.single_task<ReassociateOnKernel>([=]() [[intel::kernel_args_restrict]] {
+#pragma clang fp reassociate(on)
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/src/fp_pragmas.cpp]: Please add comment here explaining what this pragma allows the compiler to do to the floating point operations below.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/src/fp_pragmas.cpp [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+#endif
+
+  // Test fp contract pragma
+  disable_contract(selector, a, b, c, d, res_contract_off);
+  enable_contract(selector, a, b, c, d, res_contract_fast);
+  // Test fp reassociate pragma
+  disable_reassociate(selector, a, b, c, d, res_reassociate_off);
+  enable_reassociate(selector, a, b, c, d, res_reassociate_on);
+
+  if (fabs(res_contract_off - golden_res_contract) <= EPSILON &&
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/src/fp_pragmas.cpp]: Please add a comment here explaining how the floating point results will differ because of the pragmas, but they should still be 'close' to the strict results.

**Review by mdbtucker — COMMENTED** [review]
Thanks for creating this tutorial Cassie!

I feel like you need some more comments in your code, and I'm concerned that the readme doesn't do a good enough job explaining when a user would use these pragmas and what the effect would be.  If you need help with the revisions please ask Ajay or someone on your team.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/src/fp_pragmas.cpp [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+    event e = q.submit([&](handler &h) {
+      accessor accessorA(bufferA, h, read_only);
+      accessor accessorB(bufferB, h, read_only);
+      accessor accessorC(bufferC, h, read_only);
+      accessor accessorD(bufferD, h, read_only);
+      accessor accessorRes(bufferE, h, write_only, noinit);
+
+      // FPGA-optimized kernel
+      // Using kernel_args_restrict tells the compiler that the input
+      // and output buffers won't alias.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/src/fp_pragmas.cpp]: Please use this comment (taken from the fpga_compile tutorial) with appropriate modifications here and everywhere you use kernel_args_restrict:

        // The "kernel_args_restrict" tells the compiler that a, b, and r
        // do not alias. For a full explanation, see:
        //    DPC++FPGA/Tutorials/Features/kernel_args_restrict

**Comment by JoeOster** [comment]
Please note that your fork isnt upto date and that there are two more steps to add you the description

[] Compile code using compiler flags and fix anything detected "enable /Wall -Wformat-security -Werror=format-security"
[] Review Sample with Domain Expert:

Obviously, the review is done, but will need confirmation that you have run the compiler flags and fixed any errors before submission to the oneapi-src/oneAPI-samples repo

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/src/fp_pragmas.cpp (hunk 11 above)
Comment (cassiekexin) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/src/fp_pragmas.cpp]: Yes I did

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md (hunk 4 above)
Comment (cassiekexin) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md]: yes, fp contract only applies to double precision, I should remove "mainly" here

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md (hunk 6 above)
Comment (cassiekexin) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/README.md]: Yes the default for these is fp contract(fast) and fp reassociate(on) due to fast-math setting to on by default

File: DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/sample.json (hunk 9 above)
Comment (cassiekexin) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/fp_pragmas/sample.json]: Do we need to change all the occurrence of 'fp_pragmas' to 'floating_point_pragmas'?

File: DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/README.md [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+
+```
+#pragma clang fp contract(fast)
+  double temp1 = 0.0, temp2 = 0.0;
+  temp1 = accessorA[0] + accessorB[0];
+  temp2 = accessorC[0] + accessorD[0];
+  accessorRes[0] = temp1 * temp2;
+```
+
+The `fp contract(fast)` pragma will fuse the multiply and add operations into an FMA, and therefore reduce the area.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/README.md]: You have not explained what an 'FMA' is.  How about this:

The fp contract(fast) pragma allows the compiler to skip rounding steps on these double precision operations, which in turn allows more efficient use of FPGA floating point math resources and thus a reduction in area.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/README.md [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+
+### Description of the pragmas
+The `fp contract(fast|off)` pragma controls whether the compiler can skip intermediate rounding and conversions between double precision arithmetic operations. If multiple occurrences of this pragma affect the same scope of your code, the pragma with the narrowest scope takes precedence.
+
+The `fp reassociate(on|off)` pragma controls the relaxing of the order of floating point arithmetic operations within the code block that this pragma is applied to. If multiple occurrences of this pragma affect the same scope of your code, the pragma with the narrowest scope takes precedence.
+
+#### Example
+This tutorial design applies `fp contract(off)` and `fp reassociate(off)` at a global scope.
+
+A kernel in this tutorial applies `fp contract(fast)` to addition followed by multiplication. The following is an example:
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/README.md]: Please specify which kernel (don't just say 'a kerenl')

File: DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/README.md [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+#pragma clang fp contract(fast)
+  double temp1 = 0.0, temp2 = 0.0;
+  temp1 = accessorA[0] + accessorB[0];
+  temp2 = accessorC[0] + accessorD[0];
+  accessorRes[0] = temp1 * temp2;
+```
+
+The `fp contract(fast)` pragma will fuse the multiply and add operations into an FMA, and therefore reduce the area.
+
+A kernel in this tutorial applies `fp reassociate(on)` to a sequence of addition. The following is an example:
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/README.md]: Please specify which kernel, not just 'a kernel'.

'...sequence of additionS' (plural, not singular)

File: DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/README.md [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+
+```
+#pragma clang fp reassociate(on)
+  accessorRes[0] = 
+    accessorA[0] + accessorB[0] + accessorC[0] + accessorD[0];
+```
+
+The `fp reassociate(on)` will relax the order of the series of add operations and remove the dependency of each add, and therefore reduce the area.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/README.md]: How about this:

By relaxing the order of the additions, the compiler is able to group these four additions in a way that maps more efficiently to the FPGA hardware, and thus save area.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/src/fp_pragmas.cpp [hunk 21]
```diff
@@ -0,0 +1,269 @@
...
+      accessor accessorC(bufferC, h, read_only);
+      accessor accessorD(bufferD, h, read_only);
+      accessor accessorRes(bufferE, h, write_only, noinit);
+
+      // The "kernel_args_restrict" tells the compiler that a, b, and r
+      // do not alias. For a full explanation, see:
+      //    DPC++FPGA/Tutorials/Features/kernel_args_restrict
+      h.single_task<ContractFastKernel>([=]() [[intel::kernel_args_restrict]] {
+
+      // Allows the fusing of mult and add instructions into an FMA.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/src/fp_pragmas.cpp]: How about:

// This pragma allows re-ordering of the multiply and add instructions to an order that maps more efficiently into FPGA hardware

File: DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/src/fp_pragmas.cpp [hunk 22]
```diff
@@ -0,0 +1,269 @@
...
+      accessor accessorC(bufferC, h, read_only);
+      accessor accessorD(bufferD, h, read_only);
+      accessor accessorRes(bufferE, h, write_only, noinit);
+
+      // The "kernel_args_restrict" tells the compiler that a, b, and r
+      // do not alias. For a full explanation, see:
+      //    DPC++FPGA/Tutorials/Features/kernel_args_restrict
+      h.single_task<ReassociateOnKernel>([=]() [[intel::kernel_args_restrict]] {
+
+      // Enables the compiler to reorder floating-point operations to improve
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/src/fp_pragmas.cpp]: // This pragma enables . . .

File: DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/src/fp_pragmas.cpp [hunk 23]
```diff
@@ -0,0 +1,269 @@
...
+  // Test fp contract pragma
+  disable_contract(selector, a, b, c, d, res_contract_off);
+  enable_contract(selector, a, b, c, d, res_contract_fast);
+  // Test fp reassociate pragma
+  disable_reassociate(selector, a, b, c, d, res_reassociate_off);
+  enable_reassociate(selector, a, b, c, d, res_reassociate_on);
+
+  // Using fp contract and fp reassociate pragmas will cause some nuances in
+  // precision due to rounding or reordering of math operations, but the results
+  // should be close to the strict results.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/src/fp_pragmas.cpp]: How about this:

The fp contract and reassociate pragmas can save area and improve performance, but because the operations can be reordered we may get slightly different results from those computed by the CPU.  The EPISILON constant allows us to define what we consider to be 'close enough' to an exact match.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/README.md [hunk 24]
```diff
@@ -0,0 +1,182 @@
+# Contraction and Reassociation on Floating Point Numbers
+This FPGA tutorial explains how to use the fp reassociate and fp contract pragmas for floating point numbers.
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/README.md]: Please add more detail here.  Here's what I recommend:

This FPGA tutorial explains how to use the fp reassociate and fp contract pragmas for floating point numbers.  These pragmas allow the compiler to change the order of certain floating point operations so that they map more efficiently to hardware.  The impact of this reordering is that the results may be altered slightly due to rounding that can occur after each floating point operation.

File: DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/README.md [hunk 25]
```diff
@@ -0,0 +1,182 @@
...
+
+
+## Purpose
+This tutorial demonstrates a simple example of using the `fp contract(fast|off)` and `fp reassociate(on|off)` pragmas.
+
+### Description of the pragmas
+The `fp contract(fast|off)` pragma controls whether the compiler can skip intermediate rounding and conversions between double precision arithmetic operations. If multiple occurrences of this pragma affect the same scope of your code, the pragma with the narrowest scope takes precedence.
+
+The `fp reassociate(on|off)` pragma controls the relaxing of the order of floating point arithmetic operations within the code block that this pragma is applied to. If multiple occurrences of this pragma affect the same scope of your code, the pragma with the narrowest scope takes precedence.
+
```
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/README.md:27]: Please pull out the 'If multiple occurances . . .' and only say this once.  So describe the two pragmas as you already do, then add a final sentence that says 'If multiple occurances of either of these pragmas affect ....'

**Review by mdbtucker — COMMENTED** [review]
General comment: this tutorial seems to show how we save area with these pragmas, can they also improve fmax and/or lower latency?  If so, should we be illustrating that as well?  Or at least mentioning it in the README?

**Comment by cassiekexin** [comment]
> General comment: this tutorial seems to show how we save area with these pragmas, can they also improve fmax and/or lower latency? If so, should we be illustrating that as well? Or at least mentioning it in the README?

I do find latency decrease but no fmax improvement. I added this observation in the README.md

**Comment by mdbtucker** [comment]
> > General comment: this tutorial seems to show how we save area with these pragmas, can they also improve fmax and/or lower latency? If so, should we be illustrating that as well? Or at least mentioning it in the README?
> 
> I do find latency decrease but no fmax improvement. I added this observation in the README.md

In your particular case there is no fmax improvement, but in the more general case is using these pragmas expected to improve fmax and/or throughput, or lower latency?  If so I think we should mention that in the README (and also mention that this possible improvement is not illustrated by this particular code).  I thought this was one of the main reasons to use this pragma?

**Comment by mdbtucker** [comment]
As discussed offline, this tutorial needs to be reworked, closing this PR.
